### DISCUSSION TITLE
/articles/microsoft-teams-port-office-365-connector-teams-bot/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#311] [Comments]
Learn how to transition from Office 365 Connectors to Microsoft Teams Bots for better security, management, and enhanced capabilities in your Teams channels.
https://www.andrewconnell.com/articles/microsoft-teams-port-office-365-connector-teams-bot/

**Comment by ferrarirosso** [comment]
I understand that Microsoft is retiring these webhooks (aka Office Connectors) because they pose some security risks and cannot effectively secure this type of endpoint.

### Teams Webhook

There is still the possibility to define Teams Webhooks and trigger a flow to send an adaptive card, as described here:
- [Microsoft Teams Webhook](https://learn.microsoft.com/en-us/connectors/teams/?tabs=text1%2Cdotnet#microsoft-teams-webhook)
- [Add Incoming Webhook in Teams](https://learn.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/add-incoming-webhook?tabs=newteams%2Cdotnet)
- [Create Incoming Webhooks with Workflows for Microsoft Teams](https://prod.support.services.microsoft.com/en-us/office/create-incoming-webhooks-with-workflows-for-microsoft-teams-8ae491c7-0394-4861-ba59-055e33f75498)

### Questions:

1. Is the webhook here unauthenticated?
2. In the code from the second example, I don't see any authentication steps. Could you clarify where this is described?
3. In the third link above, it mentions that you can define who can trigger the flow (unauthenticated, all tenant users, or specific tenant users). This doesn't mean the webhook is authenticated, correct?

I also found a question here with an “obscure” answer, but I don’t quite understand how the Graph API fits in with the webhook and authentication:
- [Microsoft Q&A: Add Setup Basic OAuth Authentication to Incoming Webhooks](https://learn.microsoft.com/en-us/answers/questions/1166373/how-add-setup-basic-oauth-authentication-to-incomi)

---

### Your Solution: Teams Notification Bot

The solution you presented with a Teams bot is secure and doesn’t pose security risks, as it always operates in an authenticated context (assuming the app registration is done with the correct settings), right?

---

### Webhooks in General

Julie Turner had an example about Webhooks on Lists / Doclibs:
- [Julie Turner's Public Webhooks Samples](https://github.com/juliemturner/Public-Samples/tree/main/Webhooks)

Are these kinds of webhooks also unauthenticated?

**Reply by andrewconnell** [reply]
First, the only thing that's being retired are Office 365 Connectors, which is just an installable Microsoft Teams incoming webhook. The retirement scope, so far, hasn't included Microsoft Teams incoming & outgoing webhooks.

*But, IMHO, it's only a matter of time before those are retired too. Microsoft has been advising against using them for many years.*

Let me answer the rest of your questions 1-by-1...

> @ferrarirosso said: "Is the webhook here unauthenticated?"

MS Teams incoming & outgoing webhooks (including Office 365 Connectors) aren't authenticated in any way. Outgoing webhooks do include an [HMAC](https://security.stackexchange.com/questions/20129/how-and-when-do-i-use-hmac/20301), but that's not authentication. That's just a way to digitally sign the payload so the receiver, your endpoint, can verify only Teams send the HTTP POST to your endpoint. It's sort of like a public-private key pair, except it's just a string-based hash way to digitally sign the payload.

That's why I say there's no authentication with outgoing webhooks - hashing isn't authentication.

> @ferrarirosso said: "In the code from the second example, I don't see any authentication steps. Could you clarify where this is described?"

I assume you're referring to the [link you shared for an incoming webhook](https://learn.microsoft.com/microsoftteams/platform/webhooks-and-connectors/how-to/add-incoming-webhook?tabs=newteams%2Cdotnet)... in that case, there is no authentication. The only thing that "secures" the incoming webhook is the obfuscated endpoint Teams shares as part of the registration.

That's why I say there's no authentication with incoming webhooks.

> @ferrarirosso said: "In the third link above, it mentions that you can define who can trigger the flow (unauthenticated, all tenant users, or specific tenant users). This doesn't mean the webhook is authenticated, correct?"

No, the workflow might be secured, but the webhook itself isn't. Just to prove the point, [check out this video I created for Microsoft years ago](https://youtu.be/amvh4rzTCS0?si=Wr_oX7lQO7TJ86zY&t=99). At 1:39, I show how to take the URL for the incoming webhook and submit an HTTP POST with just a payload to the endpoint... no `Authentication` HTTP header or anything special in the URL... Teams accepts it.

> @ferrarirosso said: "The solution you presented with a Teams bot is secure and doesn’t pose security risks, as it always operates in an authenticated context (assuming the app registration is done with the correct settings), right?"

Correct, because the only way to communicate with Teams via a bot is via the Azure AI Bot Service, and the only way to do that is to first obtain an access token from Entra ID by authenticating as the app that's associated with the bot.

> @ferrarirosso said: "@juliemturner had an example about [Webhooks on Lists / Doclibs](https://github.com/juliemturner/Public-Samples/tree/main/Webhooks) Are these kinds of webhooks also unauthenticated?

Correct, those aren't authenticated either... that's simply you getting notified something happened by Microsoft Graph.

In general, as a tech, webhooks aren't authenticated or secure. They might be a bit more *secure* (I use that term very loosely) by signing or hashing the payload so the recipient can ensure the sender is who they expected.

Usually, you're just notified something happened. Some implementations include data, some are just informative.

## But... keep in mind - this post is about O365 Connectors (and incoming webhooks)

There's no way to secure these. Incoming webhooks allow you to post a message to a channel. The secure ways to do that:

1. use a bot (which can only be sent via an authenticated HTTP POST to the Azure AI Bot Service) or
1. using the Microsoft Graph's teamwork endpoint, which must also be sent via an authenticated HTTP POST to the Graph endpoint

**Reply by ferrarirosso** [reply]
@andrewconnell
Great thanks ! that's a fantastic answer !!!
